File: src/api/metadata.ts
// api/metadata.ts —— 元数据查询命令（无缓存，缓存由 metadataStore 负责）

import { invoke } from './client';
import type { ConnectionProfile } from '../types/connection';
import type {
  TableDetail,
  ViewDetail,
  FunctionDetail,
  RoutineDetail,
  RoutineParamInfo,
  MetadataRecord,
} from '../types/metadata';

export const metadataApi = {
  listDatabases: (profile: ConnectionProfile): Promise<string[]> =>
    invoke<string[]>('metadata_list_databases', { profile }),

  listTables: (profile: ConnectionProfile, database: string): Promise<string[]> =>
    invoke<string[]>('metadata_list_tables', { profile, schema: database }),

  listTableDetails: (
    profile: ConnectionProfile,
    database: string,
  ): Promise<TableDetail[]> =>
    invoke<TableDetail[]>('metadata_list_table_details', { profile, schema: database }),

  listViews: (profile: ConnectionProfile, database: string): Promise<string[]> =>
    invoke<string[]>('metadata_list_views', { profile, schema: database }),

  listViewDetails: (
    profile: ConnectionProfile,
    database: string,
  ): Promise<ViewDetail[]> =>
    invoke<ViewDetail[]>('metadata_list_view_details', { profile, schema: database }),

  listFunctions: (profile: ConnectionProfile, database: string): Promise<string[]> =>
    invoke<string[]>('metadata_list_functions', { profile, schema: database }),

  listRoutinesWithDetails: (
    profile: ConnectionProfile,
    database: string,
  ): Promise<RoutineDetail[]> =>
    invoke<RoutineDetail[]>('metadata_list_routines_with_details', {
      profile,
      schema: database,
    }),

  listFunctionDetails: (
    profile: ConnectionProfile,
    database: string,
  ): Promise<FunctionDetail[]> =>
    invoke<FunctionDetail[]>('metadata_list_function_details', {
      profile,
      schema: database,
    }),

  getFunctionDdl: (
    profile: ConnectionProfile,
    database: string,
    name: string,
    routineType: string,
  ): Promise<string> =>
    invoke<string>('metadata_get_function_ddl', {
      profile,
      schema: database,
      name,
      routineType,
    }),

  getRoutineParams: (
    profile: ConnectionProfile,
    database: string,
    name: string,
  ): Promise<RoutineParamInfo[]> =>
    invoke<RoutineParamInfo[]>('metadata_get_routine_params', {
      profile,
      schema: database,
      name,
    }),

  listColumns: (
    profile: ConnectionProfile,
    database: string,
    table: string,
  ): Promise<MetadataRecord[]> =>
    invoke<MetadataRecord[]>('metadata_list_columns', {
      profile,
      schema: database,
      table,
    }),

  listForeignKeys: (
    profile: ConnectionProfile,
    database: string,
    table: string,
  ): Promise<MetadataRecord[]> =>
    invoke<MetadataRecord[]>('metadata_list_foreign_keys', {
      profile,
      schema: database,
      table,
    }),

  listIndexes: (
    profile: ConnectionProfile,
    database: string,
    table: string,
  ): Promise<MetadataRecord[]> =>
    invoke<MetadataRecord[]>('metadata_list_indexes', {
      profile,
      schema: database,
      table,
    }),

  listTriggers: (
    profile: ConnectionProfile,
    database: string,
    table: string,
  ): Promise<MetadataRecord[]> =>
    invoke<MetadataRecord[]>('metadata_list_triggers', {
      profile,
      schema: database,
      table,
    }),

  listChecks: (
    profile: ConnectionProfile,
    database: string,
    table: string,
  ): Promise<MetadataRecord[]> =>
    invoke<MetadataRecord[]>('metadata_list_checks', {
      profile,
      schema: database,
      table,
    }),

  loadDdl: (
    profile: ConnectionProfile,
    database: string,
    table: string,
  ): Promise<string> =>
    invoke<string>('metadata_load_ddl', { profile, schema: database, table }),

  getAllDatabases: (profile: ConnectionProfile): Promise<string[]> =>
    invoke<string[]>('metadata_get_all_databases', { profile }),
};

File: src-tauri/src/commands/metadata.rs
use tauri::State;
use crate::errors::{AppResult, AppError};
use crate::models::connection::ConnectionProfile;
use crate::models::metadata::*;
use crate::core::pool::manager::PoolRegistry;

#[tauri::command]
pub async fn metadata_list_databases(profile: ConnectionProfile, state: State<'_, PoolRegistry>) -> AppResult<Vec<String>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_databases(&handle).await
}

#[tauri::command]
pub async fn metadata_get_all_databases(profile: ConnectionProfile, state: State<'_, PoolRegistry>) -> AppResult<Vec<String>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.get_all_databases(&handle).await
}

#[tauri::command]
pub async fn metadata_list_tables(profile: ConnectionProfile, schema: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<String>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_tables(&handle, &schema).await
}

#[tauri::command]
pub async fn metadata_list_table_details(profile: ConnectionProfile, schema: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<TableDetail>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_table_details(&handle, &schema).await
}

#[tauri::command]
pub async fn metadata_list_views(profile: ConnectionProfile, schema: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<String>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_views(&handle, &schema).await
}

#[tauri::command]
pub async fn metadata_list_view_details(profile: ConnectionProfile, schema: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<ViewDetail>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_view_details(&handle, &schema).await
}

#[tauri::command]
pub async fn metadata_list_functions(profile: ConnectionProfile, schema: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<String>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_functions(&handle, &schema).await
}

#[tauri::command]
pub async fn metadata_list_function_details(profile: ConnectionProfile, schema: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<FunctionDetail>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_function_details(&handle, &schema).await
}

#[tauri::command]
pub async fn metadata_list_routines_with_details(profile: ConnectionProfile, schema: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<RoutineDetail>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_routines_with_details(&handle, &schema).await
}

#[tauri::command]
pub async fn metadata_get_function_ddl(profile: ConnectionProfile, schema: String, name: String, routine_type: String, state: State<'_, PoolRegistry>) -> AppResult<String> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.get_function_ddl(&handle, &schema, &name, &routine_type).await
}

#[tauri::command]
pub async fn metadata_get_routine_params(profile: ConnectionProfile, schema: String, name: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<RoutineParam>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.get_routine_params(&handle, &schema, &name).await
}

#[tauri::command]
pub async fn metadata_list_columns(profile: ConnectionProfile, schema: String, table: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<ColumnInfo>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_columns(&handle, &schema, &table).await
}

#[tauri::command]
pub async fn metadata_list_foreign_keys(profile: ConnectionProfile, schema: String, table: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<ForeignKeyInfo>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_foreign_keys(&handle, &schema, &table).await
}

#[tauri::command]
pub async fn metadata_list_indexes(profile: ConnectionProfile, schema: String, table: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<IndexInfo>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_indexes(&handle, &schema, &table).await
}

#[tauri::command]
pub async fn metadata_list_triggers(profile: ConnectionProfile, schema: String, table: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<TriggerInfo>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_triggers(&handle, &schema, &table).await
}

#[tauri::command]
pub async fn metadata_list_checks(profile: ConnectionProfile, schema: String, table: String, state: State<'_, PoolRegistry>) -> AppResult<Vec<CheckInfo>> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.list_checks(&handle, &schema, &table).await
}

#[tauri::command]
pub async fn metadata_load_ddl(profile: ConnectionProfile, schema: String, table: String, state: State<'_, PoolRegistry>) -> AppResult<String> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.load_ddl(&handle, &schema, &table).await
}

#[tauri::command]
pub async fn metadata_get_current_user_info(profile: ConnectionProfile, state: State<'_, PoolRegistry>) -> AppResult<String> {
    let pool_id = state.get_or_create_pool(&profile).await?;
    let (adapter, handle) = state.get_pool_info(pool_id).ok_or(AppError::PoolNotFound(pool_id))?;
    adapter.get_current_user_info(&handle).await
}
